lib/linkerd2-opencensus/src/lib.rs

```diff
@@ -0,0 +1,180 @@
+use futures::{task, try_ready, Async, Future, Poll, Stream};
+use linkerd2_error::Error;
+use opencensus_proto::agent::common::v1::Node;
+use opencensus_proto::agent::trace::v1::{
+    client::TraceService, ExportTraceServiceRequest, ExportTraceServiceResponse,
+};
+use opencensus_proto::trace::v1::Span;
+use tokio::sync::mpsc;
+use tower_grpc::{
+    self as grpc, client::streaming::ResponseFuture, generic::client::GrpcService, BoxBody,
+};
+use tracing::trace;
+
+/// SpanExporter sends a Stream of spans to the given TraceService gRPC service.
+pub struct SpanExporter<T, S>
+where
+    T: GrpcService<BoxBody>,
+{
+    client: T,
+    node: Node,
+    state: State<T>,
+    spans: S,
+    max_batch_size: usize,
+}
+
+enum State<T: GrpcService<BoxBody>> {
+    Idle,
+    Sending {
+        sender: mpsc::Sender<ExportTraceServiceRequest>,
+        // Node data should only be sent on the first message of a streaming
+        // request.
+        node: Option<Node>,
+        // We hold the response future, but never poll it.
+        _rsp: ResponseFuture<ExportTraceServiceResponse, T::Future>,
```

Comment:
Yeah, I'm not really sure of the contract here.  If the remote sends a reset, I would expect sends to fail, even if we're not polling the response.  And I don't know the semantics of an EOS from the server.  Does the server sending a response mean that we should stop streaming request messages?

Reply:
The send stream will fail the next time you actually try to write it to, but because buffering may happen above this, it gets complicated. Polling the response side is the only way to get actively notified when the stream is torn down, afaik.